File: DSAlgorithms/Leetcode/ts/src/2_pointers/42-trapping_rain.ts
const trap  = (height: number[]): number  =>{
    if(height.length === 0)
        return 0;

    let left = 0;
    let right = height.length - 1;
    let left_max = height[left];
    let right_max = height[right];
    let sum = 0;
    while(left < right)
    {
        if(left_max < right_max)
        {
            left++;
            left_max = Math.max(left_max, height[left]);
            sum += left_max - height[left];
        }
        else
        {
            right--;
            right_max = Math.max(right_max, height[right]);
            sum += right_max - height[right];
        }
    }
    return sum;

};

File: DSAlgorithms/Leetcode/2_pointers/assorted/c++/42-trapping_rain.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if(height.size() == 0)
            return 0;
        int left = 0;
        int right = height.size() - 1;
        int left_max = height[left];
        int right_max = height[right];
        int sum = 0;
        while(left < right)
        {
            if(left_max < right_max)
            {
                left++;
                left_max = std::max(left_max, height[left]);
                sum += left_max - height[left];
            }
            else
            {
                right--;
                right_max = std::max(right_max, height[right]);
                sum += right_max - height[right];
            }
        }
        return sum;
    }
    
};